agent/consul/state/config_entry_exported_services.go

```diff
@@ -64,3 +66,76 @@ func getExportedServicesConfigEntryTxn(
 	}
 	return idx, export, nil
 }
+
+// ResolvedExportedServices returns the list of exported services along with consumers.
+// Sameness Groups and wild card entries are resolved.
+func (s *Store) ResolvedExportedServices(ws memdb.WatchSet, entMeta *acl.EnterpriseMeta) (uint64, []*pbconfigentry.ResolvedExportedService, error) {
+	tx := s.db.ReadTxn()
+	defer tx.Abort()
+
+	return resolvedExportedServicesTxn(tx, ws, entMeta)
+}
+
+func resolvedExportedServicesTxn(tx ReadTxn, ws memdb.WatchSet, entMeta *acl.EnterpriseMeta) (uint64, []*pbconfigentry.ResolvedExportedService, error) {
+	var resp []*pbconfigentry.ResolvedExportedService
+
+	// getSimplifiedExportedServices resolves the sameness group information to partitions and peers.
+	maxIdx, exports, err := getSimplifiedExportedServices(tx, ws, nil, *entMeta)
+	if err != nil {
+		return 0, nil, err
+	}
+	if exports == nil {
+		return maxIdx, nil, nil
+	}
+
+	// Services -> ServiceConsumers
+	var exportedServices = make(map[structs.ServiceName]map[structs.ServiceConsumer]struct{})
+
+	// Helper function for inserting data and auto-creating maps.
+	insertEntry := func(m map[structs.ServiceName]map[structs.ServiceConsumer]struct{}, service structs.ServiceName, consumers []structs.ServiceConsumer) {
+		for _, c := range consumers {
+			cons, ok := m[service]
+			if !ok {
+				cons = make(map[structs.ServiceConsumer]struct{})
+				m[service] = cons
+			}
+
+			cons[c] = struct{}{}
+		}
+	}
+
+	for _, svc := range exports.Services {
+		// Prevent exporting the "consul" service.
+		if svc.Name == structs.ConsulServiceName {
+			continue
+		}
+
+		svcEntMeta := acl.NewEnterpriseMetaWithPartition(entMeta.PartitionOrDefault(), svc.Namespace)
+		svcName := structs.NewServiceName(svc.Name, &svcEntMeta)
+
+		// If this isn't a wildcard, we can simply add it to the list of services to watch and move to the next entry.
+		if svc.Name != structs.WildcardSpecifier {
+			insertEntry(exportedServices, svcName, svc.Consumers)
```

Comment:
Instead of converting to an intermediate map and then to another response why not add the list of services encountered here to a slice and pass it on to the `prepareExportedServicesResponse` function? There you could just walkover the list of services and compute the response as needed.

Reply:
The main purpose of the map is to avoid duplicate entries for service and consumers. Are you suggesting to create the map in the `prepareExportedServicesResponse` function?

Reply:
Yes. The map is still needed but it just looks cleaner if there a function that accepts a list of exported services and returns back the expected `proto` response. The function's signature doesn't look that readable as of now. But I would leave it upto your best judgement here and let you take a call.

Reply:
Moved the duplicate removal logic to a new function.